cinema-model/src/main/java/touk/recru/app/entity/BaseEntity.java

```diff
@@ -0,0 +1,54 @@
+package touk.recru.app.entity;
+
+import jakarta.persistence.*;
+import lombok.*;
+import lombok.experimental.SuperBuilder;
+import org.hibernate.Hibernate;
+import org.hibernate.annotations.CreationTimestamp;
+import org.hibernate.annotations.UpdateTimestamp;
+
+import java.time.LocalDateTime;
+import java.util.Objects;
+import java.util.UUID;
+
+@MappedSuperclass
+@Setter
+@Getter
+@SuperBuilder
+@NoArgsConstructor
+@ToString
+public abstract class BaseEntity {
+	@Id
+	@GeneratedValue(strategy = GenerationType.SEQUENCE)
+	private Long id;
+
+	@Column(
+			name = "uuid",
+			columnDefinition = "uuid",
+			updatable = false,
+			nullable = false
+	)
+	@Builder.Default
+	private UUID uuid = UUID.randomUUID();
+	@CreationTimestamp
+	private LocalDateTime created;
+	@UpdateTimestamp
+	private LocalDateTime updated;
+
+	@Override
+	public int hashCode() {
+		return getClass().hashCode();
```

Comment:
Hmm, tutaj dla dwóch różnych obiektów (różne pk) z tej samej encji zostanie zwrócony ten sam hashcode. Czy to dobrze?

Reply:
Znalazłem na stacku ciekawy [artykuł](https://web.archive.org/web/20171211235806/http://www.onjava.com/pub/a/onjava/2006/09/13/dont-let-hibernate-steal-your-identity.html) na temat equals/hashCode i hibernate'a.
Jeżeli chodzi o sam hashCode, to raczej nie będzie tutaj dużego problemu z tego co rozumiem oprócz problemu związanego z performencem - HashMap/HashTable/HashSet (gorzej by było gdyby dla twóch tych samych encji zwracał inny) 
Ale prawdą jest, że lepiej zaimplementować to inaczej - chociażby właśnie przy użyciu buisnessKey. W moim wypadku niestety takiego nie possiadam (UUID nie jest zawsze tym samym dla dwóch tych samych encji jezeli encje jeszcze nie są jeszcze persistence).
Z tego co także zrozumiałem dobrze by było, abym używał buisnessKey także przy equals, aby nie było sytuacji gdzie jedna "osoba" która jest już persistence zwracała fałsz przy porównaniu z identycznym obiektem (np. ma ten sam email, ponieważ email powinien być jedny dla konta).
Mam nadzieję, że dobrze wytłumaczyłem co miałem na myśli :)

Reply:
Dla equals wykorzystujesz bussinesKey => id ;)

Reply:
Hmm, no nie do końca. W tym rzecz, że id to primary key, a dokładniej Surrogate key w tym wypadku.
BuisnessKey/[NaturalKey](https://en.wikipedia.org/wiki/Natural_key), musi mieć jakieś realne odzwierciedlenie w biznesie np. właśnie unikalny email dla jednego użytkownika. Niestety ID to tylko Key potrzebny bazie danych do unikalności rekordów, ale nie odzwierciedla żadnej logiki biznesowej. Mogę się mylić ale tak to rozumiem. 
Tu [artykuł ](http://www.agiledata.org/essays/keys.html) też z definicjami poszczególnych Key.
Tutaj też bardzo fajny artykuł moim zdaniem o problemach związanych z equals i hashCode oraz ciekawych rozwiązaniach. Niestety nie zawsze jest rozwiązanie dla każdego przypadku. Ale nawiązując do pierwszego komentarza, z artykułu też wynika, że jedynie equals powinien być uzależniony od PK, a hashCode tak naprawdę nie musi, lecz tak jak mówiłem zmniejsza performence wszystkich rzeczy opartych na hashowaniu.

> According to The java.lang.Object documentation it should be perfectly ok to always return 0 for the hashCode(). The positive effect of implementing hashCode() to return unique numbers for unique objects, is that it might increase performance. The downside is that the behavior of hashCode() must be consistent with equals(). For object a and b, if a.equals(b) is true, than a.hashCode() == b.hashCode() must be true. But if a.equals(b) returns false, a.hashCode() == b.hashCode() may still be true. Implementing hashCode() as 'return 0' meets these criteria, but it will be extremely inefficient in Hash based collection such as a HashSet or HashMap.

Z tego co zrozumiałem także po przeczytaniu paru [dyskusji](https://forum.hibernate.org/viewtopic.php?f=1&t=928172) na forum hibernate, niestety jeżeli nie mamy NaturalKey to nie ma idealnego rozwiązania na wszystkie problemy.